shared/api/project.go

```diff
@@ -0,0 +1,50 @@
+package api
+
+// ProjectsPost represents the fields of a new LXD project
+//
+// API extension: projects
+type ProjectsPost struct {
+	ProjectPut `yaml:",inline"`
+
+	Name string `json:"name" yaml:"name"`
+}
+
+// ProjectPost represents the fields required to rename a LXD project
+//
+// API extension: projects
+type ProjectPost struct {
+	Name string `json:"name" yaml:"name"`
+}
+
+// ProjectFeatures represents the list of features enabled for the project
+//
+// API extension: projects
+type ProjectFeatures struct {
+	Images   bool `json:"images" yaml:"images"`
+	Profiles bool `json:"profiles" yaml:"profiles"`
+}
+
+// ProjectPut represents the modifiable fields of a LXD project
+//
+// API extension: projects
+type ProjectPut struct {
+	Description string           `json:"description" yaml:"description"`
+	Features    *ProjectFeatures `json:"features" yaml:"features"`
```

Comment:
This looks weird to me: it seems a "denormalized" child object which is not really a child. I don't think we have this pattern anywhere else in the API, and it feels unnecessary/confusing.

I'd suggest either to drop ProjectFeatures entirely and inline these fields:

```go
type ProjectPut struct {
	Description string           `json:"description" yaml:"description"`
	HasImages bool
        HasProfiles bool
}
```

(just s/HasImages/WhateverNicer/ if you will)

Or alternatively have a true child object as list of strings:

```go
type ProjectPut struct {
	Description string           `json:"description" yaml:"description"`
	Features []string
}
```

This feels odd as well, but slightly less. My vote would be for inlining, there's no need to "namespace" attributes (project.Features.Images vs project.HasImages, or even project.FeaturesImages).

Reply:
Flatening should be fine, but we'll need logic similar to the Patch handling to detect whether the user set those fields or not during POST.

The separate struct was allowing for this:
```golang
    if project.Features == nil {
        project.Features = &api.ProjectFeatures{
            Images:   true,
            Profiles: true,
        }
    }
```

Reply:
Closing this one in favor of @freeekanayaka

Reply:
@stgraber hm I see. It seems an implementation-related quirk: if in the json structure that you post you omit something, and that something has defaults, we should be able to handle the situation without needing special tricks like this.

Reply:
I would expect the Go json decoder to have support for defaults using tags: `json:"name" default:"true"` or something. I'll look into that.

Reply:
@freeekanayaka I don't remember it having something like that, but you can use the double json decoding trick that we do in the Patch handler to check whether something was actually set in the json data.

Reply:
@stgraber It's actually very simple, you just need to set the object that you pass to json.Unmarshal() with the defaults that you want. If the users have provided values for those fields, they will be overridden, otherwise they remain untouched.

Reply:
Oh, that's neat, we should probably be doing that everywhere where we currently double parse the data.